pkg/cmd/roachtest/roachtestutil/task/manager.go

```diff
@@ -98,10 +98,22 @@ func (m *manager) GoWithCancel(fn Func, opts ...Option) context.CancelFunc {
 		}
 		err = internalFunc(l)
 		event := Event{
-			Name:           opt.Name,
-			Err:            err,
-			ExpectedCancel: err != nil && IsContextCanceled(groupCtx) && expectedContextCancellation.Load(),
+			Name: opt.Name,
+			Err:  err,
+			// TriggeredByTest is set to true if the task was canceled intentionally,
+			// by the test, and we encounter an error. The assumption is that we
+			// expect the error to have been caused by the cancelation, hence the
+			// error above was not caused by a failure. This ensures we don't register
+			// a test failure if the task was meant to be canceled. It's possible that
+			// `expectedContextCancellation` could be set before the context is
+			// canceled, thus we ensure also ensure that the context is canceled.
+			TriggeredByTest: err != nil && IsContextCanceled(groupCtx) && expectedContextCancellation.Load(),
 		}
+		// The select below intentionally does not send events if the context is
```

Comment:
If the `m.events` channel is ready _and_ the context is cancelled, select will probabilistically choose one of the options; i.e., an event may still be enqueued.

Reply:
:facepalm: Yes :100:, I've been justifying the original code : https://github.com/cockroachdb/cockroach/blob/release-24.2/pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go#L803
But maybe there's a bug in there. I'll update this to an if, because I think the original intention is still that the event should not fire if the context has already been cancelled.